zetta_utils/augmentations/imgaug.py

```diff
@@ -162,26 +170,121 @@ def imgaug_augment(
+    ref_images = None
+    for group in augmentables.values():
+        if "images" in group:
+            images_ = group["images"]
+            assert isinstance(images_, np.ndarray | torch.Tensor | Sequence)
+            # mypy seems to have a bug where checking against a union_object doesn't work
+            # e.g., the below will not pass checks
+            # a = np.ndarray | torch.Tensor | Sequence[np.ndarray | torch.Tensor]
+            # assert isinstance(images_, a)
+            ref_images = _ensure_nxyc(images_)
+            break
+    if ref_images is None:
         raise ValueError("Expected at least one image in `images` or `kwargs`")
 
-    for aug_group in augmentables.values():
-        res = augmenter.augment(
-            images=_ensure_nxyc(aug_group["images"]),  # type: ignore
-            heatmaps=_ensure_nxyc(aug_group["heatmaps"])  # type: ignore
-            if "heatmaps" in aug_group
-            else None,
-            segmentation_maps=_ensure_nxyc(aug_group["segmentation_maps"])  # type: ignore
-            if "segmentation_maps" in aug_group
-            else None,
-            keypoints=aug_group.get("keypoints", None),
-            bounding_boxes=aug_group.get("bounding_boxes", None),
-            polygons=aug_group.get("polygons", None),
-            line_strings=aug_group.get("line_strings", None),
-            return_batch=True,
-        )
+    for _, aug_group in augmentables.items():
+        if "images" in aug_group:
+            images = _ensure_nxyc(aug_group["images"])  # type: ignore
+        else:
+            images = ref_images
+
+        if mode == "2d":
+            res = augmenter.augment(
+                images=images,
+                heatmaps=_ensure_nxyc(aug_group["heatmaps"])  # type: ignore
+                if "heatmaps" in aug_group
+                else None,
+                segmentation_maps=_ensure_nxyc(aug_group["segmentation_maps"])  # type: ignore
+                if "segmentation_maps" in aug_group
+                else None,
+                keypoints=aug_group.get("keypoints", None),
+                bounding_boxes=aug_group.get("bounding_boxes", None),
+                polygons=aug_group.get("polygons", None),
+                line_strings=aug_group.get("line_strings", None),
+                return_batch=True,
+            )
+
+        elif mode == "3d":
+            # Process the stack one at a time while making sure the same transformation
+            # is being applied. See https://github.com/aleju/imgaug/issues/51
+            assert images is not None
+            assert isinstance(images, np.ndarray | torch.Tensor)
+
+            segmentation_maps = None
+            if "segmentation_maps" in aug_group:
+                assert isinstance(aug_group["segmentation_maps"], np.ndarray | torch.Tensor)
+                segmentation_maps = _ensure_nxyc(aug_group["segmentation_maps"])
+                # if segmentation_maps.dtype != np.uint8:
```

Comment:
Should this check be uncommented?

Reply:
Nah, I find while writing tests that imgaug has a check internally.